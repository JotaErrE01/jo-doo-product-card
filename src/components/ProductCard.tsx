import React, { createContext } from 'react';
import styles from '../styles/styles.module.css';
import { useProduct } from '../hooks/useProduct';
import { IProduct, ProductCardContext, OnChangeArgs, InitialValues, ProductCartHandlers } from '../interfaces/ProductsInterfaces';


export const ProductContext = createContext({ } as ProductCardContext);
const { Provider } = ProductContext;

export interface Props {
  product: IProduct;
  // children?: ReactElement | ReactElement[];
  children: ( args: ProductCartHandlers ) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  onChange?: ( args: OnChangeArgs ) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {

  const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

  return (
    <Provider value={{
      counter,
      increaseBy,
      product,
      maxCount
    }}>
      <div
        className={`${styles.productCard} ${className}`}
        style={style}
      >

        {
          children({
            count: counter,
            isMaxCountReached,
            maxCount: initialValues?.maxCount,
            product,
            increaseBy,
            reset
          })
        }

      {/* <ProductImage img={product.img} />

      <ProductTitle title={product.title} />

      <ProductButtons counter={counter} increaseBy={increaseBy} /> */}

      </div>
    </Provider>
  )
}

//estamos agregando propiedades a la interface de ProductCard, transformando el componente como si de un objeto se tratase //? OJO => RECORDAR QUE LO TODO EN JS ES UN OBJETO MENOS LOS PRIMITIVOS, POR ENDE PODEMOS AGREGAR ESTAS PROPIEDADES COMO UN OBJETO DE TODA LA VIDA Y ACCEDER A ELLAS DE IGUAL FORMA
// ProductCard.Title = ProductTitle;
// ProductCard.Image = ProductImage;
// ProductCard.Buttons = ProductButtons;
