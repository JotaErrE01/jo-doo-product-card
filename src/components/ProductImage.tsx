import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';

export interface Props {
  className?: string;
  style?: React.CSSProperties;
}

export const ProductImage = ({ className, style }: Props) => {

  const { product: { img } } = useContext(ProductContext);

  return (
    <img 
      className={`${styles.productImg} ${className}`}
      style={style}
      src={img || noImage} 
      alt="No Img"
    />
  )
}