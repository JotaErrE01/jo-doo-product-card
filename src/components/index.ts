import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';
import { ProductCard } from './ProductCard';
import { IProductCardObj } from '../interfaces/ProductsInterfaces';

//recordar que todo en js es un objeto a excepcion de los primitivos
//asign lee notion o la documentacion para entender si se te olvido, recuerda que las referencias se mantienen
const ProductCardObj: IProductCardObj = Object.assign( ProductCard, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons,
});

export default ProductCardObj;

export{
    ProductTitle,
    ProductImage,
    ProductButtons,
    ProductCardObj
}
