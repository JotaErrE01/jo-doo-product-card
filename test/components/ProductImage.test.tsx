import React from 'react';
import render from 'react-test-renderer';
import { ProductCardObj as ProductCard, ProductImage } from '../../src/components';
import { product2, product1 } from '../data/products';


describe('Product Title', () => {
    
    test('Debe de mostrar el componente con la imagen del producto', () => {
        const wrapper = render.create(
            <ProductCard
                product={ product2 }
            >
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('Debe de mostrar el componente con la imagen personalizada', () => {
        Object.assign(product1, { img: 'customimg.jpg' });
        const wrapper = render.create(
            <ProductCard
                product={ product1 }
            >
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        );

        expect(wrapper.toTree()).toMatchSnapshot();
    })  
    
})