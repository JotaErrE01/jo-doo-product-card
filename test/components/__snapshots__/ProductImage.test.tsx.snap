// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Title Debe de mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="No Img"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`Product Title Debe de mostrar el componente con la imagen personalizada 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": [Function],
    "product": Object {
      "id": "1",
      "img": "customimg.jpg",
      "title": "Coffee Mug - Card",
    },
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "host",
    "props": Object {
      "children": <ProductImage />,
      "className": "productCard undefined",
      "style": undefined,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {},
        "rendered": Object {
          "instance": null,
          "nodeType": "host",
          "props": Object {
            "alt": "No Img",
            "className": "productImg undefined",
            "src": "customimg.jpg",
            "style": undefined,
          },
          "rendered": Array [],
          "type": "img",
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  "type": [Function],
}
`;
